examples/hyper.rs

```diff
@@ -0,0 +1,81 @@
+use hyper::{
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server,
+};
+use prometheus_client::{encoding::text::encode, metrics::counter::Counter, registry::Registry};
+use std::{
+    future::Future,
+    io,
+    net::{IpAddr, Ipv4Addr, SocketAddr},
+    pin::Pin,
+    sync::Arc,
+};
+use tokio::signal::unix::{signal, SignalKind};
+
+#[tokio::main]
+async fn main() {
+    let request_counter = Counter::default();
+
+    let registry = register_metrics(&request_counter);
+
+    // Spawn a server to serve the OpenMetrics endpoint.
+    let metrics_addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8001);
+    tokio::task::spawn(start_metrics_server(metrics_addr, registry));
+}
+
+pub fn register_metrics(example_counter: &Counter) -> Registry {
+    let mut registry = <Registry>::with_prefix("tokio_hyper_example");
+
+    registry.register(
+        "requests",
+        "How many requests the application has received",
+        Box::new(example_counter.clone()),
+    );
+
+    registry
+}
+
+/// Start a HTTP server to report metrics.
+pub async fn start_metrics_server(metrics_addr: SocketAddr, registry: Registry) {
+    let mut shutdown_stream = signal(SignalKind::terminate()).unwrap();
+
+    eprintln!("Starting metrics server on {metrics_addr}");
+
+    let registry = Arc::new(registry);
+    Server::bind(&metrics_addr)
+        .serve(make_service_fn(move |_conn| {
+            let registry = registry.clone();
+            async move {
+                let handler = make_handler(registry);
+                Ok::<_, io::Error>(service_fn(handler))
+            }
+        }))
+        .with_graceful_shutdown(async move {
+            shutdown_stream.recv().await;
+        })
+        .await
+        .unwrap();
+}
+
+/// This function returns a HTTP handler (i.e. another function)
+pub fn make_handler(
+    registry: Arc<Registry>,
+) -> impl Fn(Request<Body>) -> Pin<Box<dyn Future<Output = io::Result<Response<Body>>> + Send>> {
+    // This closure accepts a request and responds with the OpenMetrics encoding of our metrics.
+    move |_req: Request<Body>| {
+        let reg = registry.clone();
+        Box::pin(async move {
+            let mut buf = Vec::new();
+            encode(&mut buf, &reg.clone()).map(|_| {
+                let body = std::str::from_utf8(buf.as_slice()).unwrap().to_string();
```

Comment:
Given that one can build a `Body` from a `[u8]` why is the conversion to a string necessary here?

https://docs.rs/hyper/latest/hyper/struct.Body.html#impl-From%3C%26%27static%20%5Bu8%5D%3E-for-Body

Reply:
Ah great point, thanks for pointing that out. Simplified it.